### PR TITLE
Adds support for nanosecond resolution file information where available (Fixes: 6283)

### PR DESCRIPTION
**Pull request by migueldeicaza**
This patch adds support to the runtime to use the nanosecond
information available on Mac and Linux on `struct stat`.
`stat.st_Xtimespec` strucutre on OSX, and `stat.st_Xtime` and
`stat.st_Xtim.tv_nsec` pair on Linux.

Additionally, this adds support in `Mono.Posix` for MacOS nanosecond stat
data, it already had Linux support before.

This is a fix for: https://github.com/mono/mono/issues/6283

### CODE REVIEW

**Comment by migueldeicaza** [comment]
The tests caught various problems when running in 32 bit mode on Linux and Mac, will have to investigate. 

The years are all wrong when running in 32 bits.

File: mono/metadata/w32file-unix.c [hunk 1]
```diff
@@ -3491,11 +3491,36 @@ mono_w32file_get_attributes_ex (const gunichar2 *name, MonoIOStat *stat)
 	/* fill stat block */
 
 	stat->attributes = _wapi_stat_to_file_attributes (utf8_name, &buf, &linkbuf);
-	stat->creation_time = (((guint64) (buf.st_mtime < buf.st_ctime ? buf.st_mtime : buf.st_ctime)) * 10 * 1000 * 1000) + 116444736000000000ULL;
-	stat->last_access_time = (((guint64) (buf.st_atime)) * 10 * 1000 * 1000) + 116444736000000000ULL;
-	stat->last_write_time = (((guint64) (buf.st_mtime)) * 10 * 1000 * 1000) + 116444736000000000ULL;
 	stat->length = (stat->attributes & FILE_ATTRIBUTE_DIRECTORY) ? 0 : buf.st_size;
 
+// Multiply seconds by this
+#define SECMULT (1000*1000*10)
```
Comment (jaykrell) [inline mono/metadata/w32file-unix.c]: Suggest ULL in there, since you said tests are failing. Just guessing.

File: mono/metadata/w32file-unix.c [hunk 2]
```diff
@@ -3491,11 +3491,36 @@ mono_w32file_get_attributes_ex (const gunichar2 *name, MonoIOStat *stat)
...
+
+// Constants to convert Unix times to the API expected by .NET and Windows
+#define CONVERT_BASE  116444736000000000ULL
+
+#if HAVE_STRUCT_STAT_ST_ATIMESPEC
+	if (buf.st_mtimespec.tv_sec < buf.st_ctimespec.tv_sec || (buf.st_mtimespec.tv_sec == buf.st_ctimespec.tv_sec && buf.st_mtimespec.tv_nsec < buf.st_ctimespec.tv_nsec))
```
Comment (jaykrell) [inline mono/metadata/w32file-unix.c]: Maybe a function UnixTimeNanoSecondsToNtTime? And UnixTimeSecondsToNtTime? Or SecondsSince1980ToNtTime and NanoSecondsSince1980ToNtTime?

File: support/sys-stat.c [hunk 3]
```diff
@@ -51,16 +51,24 @@ Mono_Posix_FromStat (struct Mono_Posix_Stat *from, void *_to)
...
+	to->st_mtimespec.tv_sec = from->st_mtime_;
+	to->st_mtimespec.tv_nsec = from->st_mtime_nsec;
+	to->st_ctimespec.tv_sec = from->st_ctime_;
+	to->st_ctimespec.tv_nsec = from->st_ctime_nsec;
+#else
+#    ifdef HAVE_STRUCT_STAT_ST_ATIM
 	to->st_atim.tv_nsec = from->st_atime_nsec;
-#endif
-#ifdef HAVE_STRUCT_STAT_ST_MTIM
+#    endif
```
Comment (luhenry) [inline support/sys-stat.c]: We do not have spaces in between `#` and `if/else/endif`

Comment (migueldeicaza) [reply support/sys-stat.c]: We should, someone might have undone code that historically did that that, but it is a bad practice.

**Comment by migueldeicaza** [comment]
Jay correctly diagnosed the problem and the tests pass.   The failing tests are not related to this issue.

**Pull request merged**